---
bibliography: C:/Users/Jeff/Documents/Literature/library.bib
---


# Theory


Below are explanations of the assumptions and theory that underlie the simulations in each of the tabs.

### Randomly mating populations

This simulation tracks the frequency of the $A$ allele over generations in one or more populations of randomly mating, diploid individuals. The simulation initiates with a starting allele frequency of $p_A$. Realizations of the number of $A$ alleles in each individual are drawn from a bionomial distribution of size $N$ and success probability $p_A$, or $x_{0i} \sim \textrm{binom}(N, p_{A0})$, where $g = 0$ indicates the generation number ($0$ for the base generation) and $i$ indicates the individual number.

The mean of this realization is the average number of $A$ alleles in an individual. This estimate, divided by the number of alleles in an individual (2 for a diploid) gives the estimated frequency of $A$ alleles in the population, or $\hat{p}_{A0} = \frac{1}{2} \sum{x_{0i}}$.

We assume that each individual is equally likely to pass alleles on to the next generation. Therefore, the realization of the number of $A$ alleles in the next generation is drawn from a similar binomial distribution, except the success probability is changed to the estimated allele frequency ($\hat{p}_A$) in the previous generation, or $x_{gi} \sim \textrm{binom}(N, \hat{p}_A)$. This process is repeated for $g = 1, 2, ..., G$ generations.

The whole simulation can be summarized as follows:

1. Set the starting frequency of $A$ alleles, or $p_{A0}$
2. Sample realizations of the number of $A$ alleles, or $x_{0i} \sim \textrm{binom}(N, p_{A0})$
3. Estimate the frequency of $A$ alleles from the sample, or $\hat{p}_{A0} = \frac{1}{2} \sum{x_{0i}}$
4. For $g = 1, 2, ..., G$ generations:
    + Sample realizations of the number of $A$ alleles using the estimated allele frequency in the previous generation, or $x_{gi} \sim \textrm{binom}(N, \hat{p}_{A(g-1)})$
    + Estimate the frequency of $A$ alleles from the sample, or $\hat{p}_{Ag} = \frac{1}{2} \sum{x_{gi}}$


### Genetic variance

The genetic variance simulation relies on the theoretical equation for the additive genetic variance and dominance genetic variance at a single gene [@Bernardo2010]:

$V_A = 2pq[a+d(q-p)]^2 \\ V_D = 4p^2q^2d^2$,

where $p$ is the frequency of the $A$ allele, $q$ is the frequency of the $B$ allele ($q = 1 - p$), $a$ is the additive effect of the gene, and $d$ is the dominance effect at the gene.

Since we only consider a single gene, epistatic variance (i.e. $V_I$) is absent. Therefore, the total genetic variance at the single gene is the sum of the additive variance and the dominance variance, or $V_G = V_A + V_D$.

By using these equations, we assume that:

+ The population is in Hardy-Weinberg equalibrium
+ There are only two alleles at the gene
    

### Response to selection

This simulations combines the basic elements of the first (namely, a randomly mating population) and the second (namely, additive effects at genes). This simulation is expanded to included many genes and the dimension of selection.

As before, the number of $A$ alleles at each gene is simulated from a binomial distribution given a starting allele frequency that is identitical for all genes. The number of genes influencing the quantitative trait (i.e. QTL) is assumed to be $L = 25$. The genes are assumed independent (i.e. no linkage) and dominance is assumed absent.

The additive effects of genes are drawn from a geometric series. The effect of the *k*th QTL (where $k = 1, 2, ..., L$) is equal to $b^k$, where $b = \frac{L - 1}{L + 1}$. Greater additive effects are more favorable, and the $A$ allele is always assumed to be the favorable allele.

The genotypic value of an individual is the sum of the additive allele effects carried by that individual. The genetic variance ($V_G$) is the the variance among these genotypic values. To generate phenotypic values, residuals are sampled from a normal distribution with mean $0$ and standard deviation $\sqrt{V_R}$ and added to the genotypic values. $V_R$ is the residual (i.e. error) variance and is calculated using the user-defined heritability ($h^2$), where $V_R = \frac{V_G}{h^2} - V_G$. 

Selection is simulated by taking the top ($i \times 100$)th percent of the individuals based on phenotypic values. The estimated allele frequencies in the selected population are used to generate the allele numbers in the next generation. 



### References


